tephra-core/src/main/java/co/cask/tephra/persist/TransactionEdit.java

```diff
@@ -147,60 +161,78 @@ public static TransactionEdit createCommitting(long writePointer, Set<ChangeId>
    */
   public static TransactionEdit createCommitted(long writePointer, Set<ChangeId> changes, long nextWritePointer,
                                                 boolean canCommit) {
-    return new TransactionEdit(writePointer, 0L, State.COMMITTED, 0L, changes, nextWritePointer, canCommit);
+    return new TransactionEdit(writePointer, 0L, State.COMMITTED, 0L, changes, nextWritePointer, canCommit, null);
   }
 
   /**
    * Creates a new instance in the {@link State#ABORTED} state.
    */
   public static TransactionEdit createAborted(long writePointer) {
-    return new TransactionEdit(writePointer, 0L, State.ABORTED, 0L, null, 0L, false);
+    return new TransactionEdit(writePointer, 0L, State.ABORTED, 0L, null, 0L, false, null);
   }
 
   /**
    * Creates a new instance in the {@link State#INVALID} state.
    */
   public static TransactionEdit createInvalid(long writePointer) {
-    return new TransactionEdit(writePointer, 0L, State.INVALID, 0L, null, 0L, false);
+    return new TransactionEdit(writePointer, 0L, State.INVALID, 0L, null, 0L, false, null);
   }
 
   /**
    * Creates a new instance in the {@link State#MOVE_WATERMARK} state.
    */
   public static TransactionEdit createMoveWatermark(long writePointer) {
-    return new TransactionEdit(writePointer, 0L, State.MOVE_WATERMARK, 0L, null, 0L, false);
+    return new TransactionEdit(writePointer, 0L, State.MOVE_WATERMARK, 0L, null, 0L, false, null);
   }
 
   @Override
   public void write(DataOutput out) throws IOException {
-    CODEC_V2.encode(this, out);
+    CODEC_V3.encode(this, out);
   }
 
   @Override
   public void readFields(DataInput in) throws IOException {
     byte version = in.readByte();
-    if (V2 == version) {
-      CODEC_V2.decode(this, in);
-    } else if (V1 == version) {
-      CODEC_V1.decode(this, in);
-    } else {
-      throw new IOException("Unexpected version for edit: " + version);
+    switch (version) {
+      case V3:
+        CODEC_V3.decode(this, in);
+        break;
+      case V2:
+        CODEC_V2.decode(this, in);
+        break;
+      case V1:
+        CODEC_V1.decode(this, in);
+        break;
+      default:
+        throw new IOException("Unexpected version for edit: " + version);
     }
   }
 
   @Override
-  public boolean equals(Object obj) {
-    if (!(obj instanceof TransactionEdit)) {
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
```

Comment:
Unless the derived class also makes the same instanceof checks, the equals method might become non-symmetric when instanceof is used, right?

In our case, TransactionEditV2 does not override equals method, hence instanceof will work. I will make equals final, and use insanceof check.